### ISSUE TITLE
sample! without replacement allocates memory

### ISSUE THREAD

**Issue by tlnagy** [JuliaStats/StatsBase.jl#175]
Is this the intended behavior?
```{.julia}
x = collect(1:10^6)
y = Array(Int64, size(x))
@allocated sample!(x, y, replace=false)
```
gives `8000144` while sampling with replacement behaves as expected
```{.julia}
@allocated sample!(x, y)
```
gives `0`

**Comment by tlnagy** [comment]
It looks like it may be an unavoidable O(n) allocation needed for the Fisher-Yates shuffle

**Comment by johnmyleswhite** [comment]
FWIW, I don't see how any algorithm for sampling without replacement could avoid allocating memory since you need to remember which outcomes you've already drawn.

**Comment by tlnagy** [comment]
Yeah, I realized the same thing. I guess it would be nice to pass through a preallocated array to be used to store previous outcomes. I'm using this in a simulation where I have repeated call `sample!` on arrays with 10^7-10^8 elements and the memory pressure is causing problems.

**Comment by johnmyleswhite** [comment]
That sounds like a good idea to me.
